metrics/prometheus.go

```diff
@@ -1771,50 +1779,52 @@ func NewPrometheusCollector(i infoProvider, f ContainerLabelsFunc, includedMetri
 	return c
 }
 
-var (
-	versionInfoDesc = prometheus.NewDesc("cadvisor_version_info", "A metric with a constant '1' value labeled by kernel version, OS version, docker version, cadvisor version & cadvisor revision.", []string{"kernelVersion", "osVersion", "dockerVersion", "cadvisorVersion", "cadvisorRevision"}, nil)
-	startTimeDesc   = prometheus.NewDesc("container_start_time_seconds", "Start time of the container since unix epoch in seconds.", nil, nil)
-	cpuPeriodDesc   = prometheus.NewDesc("container_spec_cpu_period", "CPU period of the container.", nil, nil)
-	cpuQuotaDesc    = prometheus.NewDesc("container_spec_cpu_quota", "CPU quota of the container.", nil, nil)
-	cpuSharesDesc   = prometheus.NewDesc("container_spec_cpu_shares", "CPU share of the container.", nil, nil)
-)
+func (c *PrometheusCollector) SetOpts(opts v2.RequestOptions) {
+	c.opts = opts
+}
+
+//// Describe describes all the metrics ever exported by cadvisor. It
+//// implements prometheus.PrometheusCollector.
+//func (c *PrometheusCollector) Describe(ch chan<- *prometheus.Desc) {
+//	panic("unpexected")
+//	//c.errors.Describe(ch)
+//	for _, cm := range c.containerMetrics {
+//		ch <- cm.desc(nil, []string{})
+//	}
+//}
+
+// Collect fetches latest statistics about containers and exposes them to Prometheus handler that converts them
+// to expected Prometheus format.
+// Based on option it
+// * ask the latest from provided services.
+// * returns cached results if until time expires.
+//
+// It implements prometheus.rawCollector.
+func (c *PrometheusCollector) Collect() []*dto.MetricFamily {
+	if c.opts.MaxAge == nil || time.Since(c.lastUpdate) > *c.opts.MaxAge {
+		c.lastUpdate = time.Now()
+		session := c.cache.NewSession()
+
+		errorsGauge := 0
+		if err := c.collectVersionInfo(session); err != nil {
+			errorsGauge = 1
+			klog.Warningf("Couldn't get version info: %s", err)
+		}
+		if err := c.collectContainersInfo(session); err != nil {
+			errorsGauge = 1
+			klog.Warningf("Couldn't get containers: %s", err)
+		}
 
-// Describe describes all the metrics ever exported by cadvisor. It
-// implements prometheus.PrometheusCollector.
-func (c *PrometheusCollector) Describe(ch chan<- *prometheus.Desc) {
-	c.errors.Describe(ch)
-	for _, cm := range c.containerMetrics {
-		ch <- cm.desc([]string{})
+		session.MustAddMetric(
+			"container_scrape_error", "1 if there was an error while getting container metrics, 0 otherwise",
+			nil, nil, prometheus.GaugeValue, float64(errorsGauge), nil,
```

Comment:
Spent some time today trying to get this working properly in k8s.

I am finding that I keep running into errors... If I don't set `IdType: cadvisorv2.TypeName` in the cadvisor options, I get

```
ehashman@fedora:~/src/k8s$ curl http://127.0.0.1:8001/api/v1/nodes/127.0.0.1/proxy/metrics/cadvisor
# HELP cadvisor_version_info A metric with a constant '1' value labeled by kernel version, OS version, docker version, cadvisor version & cadvisor revision.
# TYPE cadvisor_version_info gauge
cadvisor_version_info{cadvisorRevision="",cadvisorVersion="",dockerVersion="",kernelVersion="5.10.0-9-amd64",osVersion="Debian GNU/Linux 11 (bullseye)"} 1
# HELP container_scrape_error 1 if there was an error while getting container metrics, 0 otherwise
# TYPE container_scrape_error gauge
container_scrape_error 1
```

and error in the logs
```
W1124 16:45:01.711801 1279370 prometheus.go:1815] Couldn't get containers: invalid request type ""
```

If I do set it, I get

```
ehashman@fedora:~/src/k8s$ curl http://127.0.0.1:8001/api/v1/nodes/127.0.0.1/proxy/metrics/cadvisor
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "error trying to reach service: EOF",
  "reason": "ServiceUnavailable",
  "code": 503
}
```

Not sure what I'm doing wrong.

k8s WIP PR is https://github.com/kubernetes/kubernetes/pull/106334

And the patch I'm applying on top of the PR to get to that second error case is

```diff
diff --git a/pkg/kubelet/server/server.go b/pkg/kubelet/server/server.go
index ce3fc942e49..be9cb5e2603 100644
--- a/pkg/kubelet/server/server.go
+++ b/pkg/kubelet/server/server.go
@@ -378,7 +378,16 @@ func (s *Server) InstallDefaultHandlers() {
                includedMetrics[cadvisormetrics.AcceleratorUsageMetrics] = struct{}{}
        }
 
-       br.MustRegisterRaw(metrics.NewPrometheusCollector(prometheusHostAdapter{s.host}, containerPrometheusLabelsFunc(s.host), includedMetrics, clock.RealClock{}))
+       cadvisorOpts := cadvisorv2.RequestOptions{
+               IdType:    cadvisorv2.TypeName,
+               Count:     1,
+               Recursive: true,
+       }
+
+       containersCollector := metrics.NewPrometheusCollector(prometheusHostAdapter{s.host}, containerPrometheusLabelsFunc(s.host), includedMetrics, clock.RealClock{})
+       containersCollector.SetOpts(cadvisorOpts)
+       br.MustRegisterRaw(containersCollector)
+
        br.MustRegister(metrics.NewPrometheusMachineCollector(prometheusHostAdapter{s.host}, includedMetrics))
        s.restfulCont.Handle(cadvisorMetricsPath,
                promhttp.HandlerForTransactional(br, promhttp.HandlerOpts{ErrorHandling: promhttp.ContinueOnError}),

```